integrations/gitlab2/gitlabcore/client.py

```diff
@@ -0,0 +1,253 @@
+        self.http_client.headers.update()
+        self.app_host = app_host
+
+    @property
+    def api_auth_header(self) -> dict[str, Any]:
+        return {
+            "Authorization": f"Bearer {self.token}"
+        }
+
+    async def _get_single_resource(self, url: str, query_params: Optional[dict[str, Any]] = None) -> dict[str, Any]:
+        retries = 0
+        while True:
+            try:
+                response = await self.http_client.get(url=url, params=query_params, headers=self.api_auth_header)
+                response.raise_for_status()
+                return response.json()
+            except httpx.HTTPStatusError as e:
+                headers = e.response.headers
+                status_code = e.response.status_code
+
+                if not await handle_retry_logic(status_code, headers, retries):
+                    logger.error(f"HTTP error: {status_code}, {e.response.text}")
+                    raise
+                retries += 1
+
+    async def get_groups(self) -> AsyncGenerator[List[Dict[str, Any]], None]:
+        paginated_groups = PaginatedData(
+            url= f"{self.base_url}/groups",
+            query_data={"per_page": 5},
+            headers=self.api_auth_header
+        )
+
+        async for group_list in paginated_groups.get_paginated_data():
+            yield group_list
+
+
+    @cache_iterator_result()
+    async def get_projects(self) -> AsyncGenerator[List[Dict[str, Any]], None]:
+        """Fetch paginated projects from GitLab and return an async generator of project lists."""
+        paginated_projects = PaginatedData(
+            url=f"{self.base_url}/projects",
+            query_data={"per_page": 5},  # You can adjust per_page as needed
+            headers=self.api_auth_header  # Pass the headers for authorization
+        )
+
+        # Use async generator to yield each page of projects
+        async for project_list in paginated_projects.get_paginated_data():
+            yield project_list
+
+    async def get_merge_requests(self) -> AsyncGenerator[List[Dict[str, Any]], None]:
+        paginated_merge_requests = PaginatedData(
+            url= f"{self.base_url}/merge_requests",
+            query_data={"per_page": 5},
+            headers=self.api_auth_header
+        )
+
+        async for merge_request_list in paginated_merge_requests.get_paginated_data():
+            yield merge_request_list
+
+    async def get_issues(self) -> AsyncGenerator[List[Dict[str, Any]], None]:
+        paginated_issues = PaginatedData(
+            url= f"{self.base_url}/issues",
+            query_data={"per_page": 5},
+            headers=self.api_auth_header
+        )
+
+        async for issue_list in paginated_issues.get_paginated_data():
+            yield issue_list
+
+
+    async def create_project_webhook(self, project_id: int):
+        url = f"{self.base_url}/projects/{project_id}/hooks"
+        webhook_url = f"{self.app_host}/integration/webhook"
```

Comment:
creating webhooks per project is going to be spammy. You can look into how to create it on a group level such as group hooks. this way, all projects under the group will inherit the hook